src/TimeTrace/ExecutionHierarchy.cpp

```diff
@@ -0,0 +1,342 @@
+
+    return &entry;
+}
+
+void ExecutionHierarchy::OnInvocation(const Invocation& invocation)
+{
+    auto it = entries_.find(invocation.EventInstanceId());
+    assert(it != entries_.end());
+
+    // may not be present, as it's not available in earlier versions of the toolset
+    if (invocation.ToolPath()) {
+        it->second.Properties.try_emplace("Tool Path", ToString(invocation.ToolPath()));
+    }
+
+    it->second.Properties.try_emplace("Working Directory", ToString(invocation.WorkingDirectory()));
+    it->second.Properties.try_emplace("Tool Version", invocation.ToolVersionString());
+
+    if (invocation.EventId() == EVENT_ID_COMPILER) {
+        it->second.Name = "CL Invocation " + std::to_string(invocation.InvocationId());
+    }
+    else if (invocation.EventId() == EVENT_ID_LINKER) {
+        it->second.Name = "Link Invocation " + std::to_string(invocation.InvocationId());
+    }
+}
+
+void ExecutionHierarchy::OnFrontEndFile(const FrontEndFile& frontEndFile)
+{
+    auto it = entries_.find(frontEndFile.EventInstanceId());
+    assert(it != entries_.end());
+    it->second.Name = frontEndFile.Path();
+}
+
+void ExecutionHierarchy::OnFinishFunction(const Activity& parent, const Function& function)
+{
+    // filter by duration
+    auto durationMs = std::chrono::duration_cast<std::chrono::milliseconds>(function.Duration());
+    if (durationMs < filter_.IgnoreFunctionUnderMs) {
+        IgnoreEntry(function.EventInstanceId(), parent.EventInstanceId());
+    }
+    else
+    {
+        auto it = entries_.find(function.EventInstanceId());
+        assert(it != entries_.end());
+        it->second.Name = function.Name();
+    }
+}
+
+void ExecutionHierarchy::OnFinishRootTemplateInstantiation(const Activity& parent, const TemplateInstantiation& templateInstantiation)
+{
+    if (!filter_.AnalyzeTemplates) {
+        IgnoreEntry(templateInstantiation.EventInstanceId(), parent.EventInstanceId());
+    }
+    else
+    {
+        // keep full hierarchy when root TemplateInstantiation passes filter, even if children wouldn't pass
+        auto rootDurationMs = std::chrono::duration_cast<std::chrono::milliseconds>(templateInstantiation.Duration());
+        if (rootDurationMs < filter_.IgnoreTemplateInstantiationUnderMs) {
+            IgnoreEntry(templateInstantiation.EventInstanceId(), parent.EventInstanceId());
+        }
+        else
+        {
+            // get us subscribed for name resolution (may already have some other activities following)
+            auto result = unresolvedTemplateInstantiationsPerSymbol_.try_emplace(templateInstantiation.SpecializationSymbolKey(),
+                                                                                 TUnresolvedTemplateInstantiationNames());
+            result.first->second.push_back(templateInstantiation.EventInstanceId());
+        }
+    }
+}
+
+void ExecutionHierarchy::OnFinishNestedTemplateInstantiation(const TemplateInstantiationGroup& templateInstantiationGroup,
```

Comment:
It looks like this function will needlessly subscribe all nested template instantiations even when the root ends up being ignored.

Wouldn't it be possible to have only 1 function like: `OnFinishTemplateInstantiation(const Activity& parent, const TemplateInstantiationGroup& templateInstantiationGroup)`? You can capture the current template instantiation with `templateInstantiationGroup.Back()` and get the parent through `parent` in case where `templateInstantiationGroup.Size() == 1`, or `templateInstantiationGroup[templateInstantiationGroup.Size() - 2]` otherwise. You can also apply the filter to the root with `templateInstantiationGroup.Front()`.

It's possible I'm missing something. Does my suggestion make sense?

Reply:
Please excuse the following wall of text.

---

In fact, that was my first approach and reasoning 🙂 

However, when I use `OnFinishTemplateInstantiation(const Activity& parent, const TemplateInstantiationGroup& templateInstantiationGroup)` these things happen:
  * `templateInstantiationGroup` always has a `Size()` of `1`.
  * `parent` can be a `TemplateInstantiation` (which means we're somewhere in the hierarchy of nested template instantiations) or activities that can be parents of `TemplateInstantiation` (which means the group only contains the root of the hierarchy).

I also tried using `OnFinishTemplateInstantiation(const TemplateInstantiationGroup& templateInstantiationGroup` (which leaves me without the parent of the root template, but let's leave that detail aside) and that captures the different sub-hierarchies from each root template as expected.

In any case, root's `StopTimestamp()` only gets assigned when it finishes. That means we've already been notified of all nested ones before we can perform the filter (if we were to check `templateInstantiationGroup.Front().Duration()` while we're processing a nested one, we'd get `0 ns`).

So, because filtering is delayed until root finishes processing, I need to keep track of all sub-hierarchies because we may pass the filter (or we may not, and we need to drop them).
That's the reason why I use two functions (one to correctly capture the root and its parent, one for sub-hierarchies, both of which are responsible of the *leaf* activity).

As for the subscriptions to `SymbolName`, I decided not to clean them when one hierarchy gets dropped and just ignore them when name gets resolved.
The reasoning was:
  * No clean-up: when notifying subscribers, for each `TemplateInstantiation` find it in `entries_`. It may not exist, but it only took one look-up to `unordered_map`.
  * Clean-up: upon filtering, for each `TemplateInstantiation`, find its subscribed key in `unresolvedTemplateInstantiationsPerSymbol_`, then `erase` from the subscription list.

I chose the former for being more simple to reason about and avoid going back and forth when cleaning up subscriptions.

Do you suggest any other way to just have one function and make it work the way you mentioned?
Also, in case I missed something, would there be any way of filtering the root without having visited all sub-hierarchies before?

Thank you!

Reply:
You're right, this is due to a bug in the SDK which does not collect all `TemplateInstantiation` activities in the group when a general `Activity` is on the left. Let's leave this like this for now.